src/sparseml/pytorch/sparsification/distillation/modifier_per_layer.py

```diff
@@ -0,0 +1,442 @@
+        else:
+            _find_layers_by_name(
+                self._teacher, self._teacher_layer_names, cached_teacher_layers
+            )
+        _LOGGER.info("Distilling teacher layers: %s", self._teacher_layer_names)
+
+        if len(self._teacher_layer_names) != len(self._student_layer_names):
+            raise ValueError(
+                "Found different numbers of teacher and student layers to distill. "
+                "Set teacher_layer_names and student_layer_names explicitly."
+            )
+
+        self._student_handles = [
+            layer.register_forward_hook(
+                _create_cache_output_hook(
+                    name, self._cached_student_output, self._student_output_shapes
+                )
+            )
+            for name, layer in cached_student_layers.items()
+        ]
+
+        self._teacher_handles = [
+            layer.register_forward_hook(
+                _create_cache_output_hook(
+                    name, self._cached_teacher_output, self._teacher_output_shapes
+                )
+            )
+            for name, layer in cached_teacher_layers.items()
+        ]
+
+    def update(
+        self,
+        module: Module,
+        optimizer: torch.optim.Optimizer,
+        epoch: float,
+        steps_per_epoch: int,
+    ):
+        super().update(module, optimizer, epoch, steps_per_epoch)
+        if epoch >= self.end_epoch:
+            for handle in self._student_handles:
+                handle.remove()
+            for handle in self._teacher_handles:
+                handle.remove()
+            self._reset_cache()
+
+    def compute_total_loss(self, loss, distillation_loss):
+        return loss + self.gain * distillation_loss
+
+    def compute_distillation_loss(self, optimizer: torch.optim.Optimizer, **kwargs):
+        if self.project_features and len(self._projection) == 0:
+            # NOTE: have to call initialize here because we need the cached output
+            # from the module. i.e. we need forward to have been called already
+            projection_params = self._initialize_projection()
+            optimizer.add_param_group({"params": projection_params})
+
+            # NOTE: since we added a param group, if someone tries
+            # to serialize the optimizer, the new param group will be
+            # included. if they try call load_state_dict() on
+            # a fresh optimizer (one that hasn't been called
+            # with compute_distillation_loss), it will
+            # crash.
+            _old_state_dict_fn = optimizer.state_dict
+
+            def state_dict_without_latest_param_group():
+                full_state_dict = _old_state_dict_fn()
+                original_state_dict = deepcopy(full_state_dict)
+                original_state_dict["param_groups"].pop()
+                return original_state_dict
+
+            optimizer.state_dict = state_dict_without_latest_param_group
```

Comment:
@bfineran this is the only way we could think of getting around the serialization issue. feels very hacky though and we aren't sure how safe this is

Reply:
I think we're going to need to override the state dict function unfortunately, don't see any way around that given this workflow. this however doesn't address the issue of reloading see my comment on reloading for that.

Reply:
additionally, on first glance, I'm not sure this properly overrides the state dict function since the method is not bound to the optimizer instance. (ie when `optimizer.state_dict()` is called, `self` will be implicitly passed in, however the function we override with doesn't include this) the test seems to pass for some reason though so maybe I am wrong here.
if we do need to bind,  I can give a quick code pointer on that